src/hotspot/share/gc/shenandoah/shenandoahGenerationalControlThread.cpp

```diff
@@ -608,159 +584,156 @@ void ShenandoahGenerationalControlThread::service_concurrent_cycle(ShenandoahHea
+      ShenandoahHeap::heap()->cancel_gc(GCCause::_shenandoah_concurrent_gc);
+      notify_cancellation(ml, GCCause::_shenandoah_concurrent_gc);
       ml.wait();
     }
     return true;
   }
 
-  if (preempt_old_marking(generation)) {
-    assert(gc_mode() == servicing_old, "Expected to be servicing old, but was: %s.", gc_mode_name(gc_mode()));
-    GCCause::Cause existing = Atomic::cmpxchg(&_requested_gc_cause, GCCause::_no_gc, GCCause::_shenandoah_concurrent_gc);
-    if (existing != GCCause::_no_gc) {
-      log_debug(gc, thread)("Reject request to interrupt old gc because another gc is pending: %s", GCCause::to_string(existing));
-      return false;
-    }
-
-    log_info(gc)("Preempting old generation mark to allow %s GC", shenandoah_generation_name(generation));
-    _requested_generation = generation;
-    _preemption_requested.set();
-    ShenandoahHeap::heap()->cancel_gc(GCCause::_shenandoah_concurrent_gc);
-    notify_control_thread();
+  if (gc_mode() == none) {
+    while (gc_mode() == none) {
+      if (_requested_gc_cause != GCCause::_no_gc) {
+        log_debug(gc, thread)("Reject request for concurrent gc because another gc is pending: %s", GCCause::to_string(_requested_gc_cause));
+        return false;
+      }
 
-    MonitorLocker ml(&_regulator_lock, Mutex::_no_safepoint_check_flag);
-    while (gc_mode() == servicing_old) {
+      notify_control_thread(ml, GCCause::_shenandoah_concurrent_gc, generation);
       ml.wait();
     }
     return true;
   }
 
+
   log_debug(gc, thread)("Reject request for concurrent gc: mode: %s, allow_old_preemption: %s",
                         gc_mode_name(gc_mode()),
                         BOOL_TO_STR(_allow_old_preemption.is_set()));
   return false;
 }
 
-void ShenandoahGenerationalControlThread::notify_control_thread() {
-  MonitorLocker locker(&_control_lock, Mutex::_no_safepoint_check_flag);
-  _control_lock.notify();
+void ShenandoahGenerationalControlThread::notify_control_thread(GCCause::Cause cause, ShenandoahGeneration* generation) {
+  MonitorLocker ml(&_control_lock, Mutex::_no_safepoint_check_flag);
+  notify_control_thread(ml, cause, generation);
+}
+
+void ShenandoahGenerationalControlThread::notify_control_thread(MonitorLocker& ml, GCCause::Cause cause, ShenandoahGeneration* generation) {
+  assert(_control_lock.is_locked(), "Request lock must be held here");
+  log_debug(gc, thread)("Notify control (%s): %s, %s", gc_mode_name(gc_mode()), GCCause::to_string(cause), generation->name());
+  _requested_gc_cause = cause;
+  _requested_generation = generation;
+  ml.notify();
+}
+
+void ShenandoahGenerationalControlThread::notify_cancellation(GCCause::Cause cause) {
+  MonitorLocker ml(&_control_lock, Mutex::_no_safepoint_check_flag);
+  notify_cancellation(ml, cause);
+}
+
+void ShenandoahGenerationalControlThread::notify_cancellation(MonitorLocker& ml, GCCause::Cause cause) {
+  assert(_heap->cancelled_gc(), "GC should already be cancelled");
```

Comment:
Not sure about this, but if cancellations can happen only when `request*` fields are clear (or may be not?), then this would be a good place to do such invariant checks (much as the suggestion above  in `notify_control_thread()`.